explorer/interpreter/type_checker.cpp

```diff
@@ -4511,7 +4511,7 @@ auto TypeChecker::DeclareClassDeclaration(Nonnull<ClassDeclaration*> class_decl,
   const int class_level = base_class ? (*base_class)->hierarchy_level() + 1 : 0;
   for (const auto* m : class_decl->members()) {
     const auto* fun = dyn_cast<FunctionDeclaration>(m);
-    if (!fun || !fun->is_virtual()) {
+    if (!fun || fun->override() == VirtualOverride::None) {
```

Comment:
Per discussion on https://github.com/carbon-language/carbon-lang/issues/2355, I think this should turn more in the direction of:

```
if (!fun) {
  continue;
}
if (fn->override() != VirtualOverride::None && !fun->is_method()) {
  cannot be virtual error
}
```

because we want it to be an error in the case of:

```
class A {
  virtual fn Foo[self]();
}
class B extends A {
  fn Foo[self]();
}
```

I'm suggesting this change here because it feels like it reinforces the `switch` code structure below. But if you'd rather split it out because it's distinct from the `impl` work, that's fine.

Reply:
Thanks for tagging 2355, added here and corresponding test.